tensorflow/contrib/avro/ops/avro_record_dataset.cc

```diff
@@ -0,0 +1,419 @@
+  AvroValueInterfacePtr p_reader_iface_;  // Reader class info to create instances
+  AvroValueInterfacePtr p_writer_iface_;  // Writer class info to create instances
+  avro_value_t reader_value_;  // Reader value, unequal from writer value when
+                               // doing schema resolution
+  avro_value_t writer_value_;  // Writer value
+};
+
+class AvroRecordDatasetOp : public DatasetOpKernel {
+ public:
+  using DatasetOpKernel::DatasetOpKernel;
+
+  void MakeDataset(OpKernelContext* ctx, DatasetBase** output) override {
+    const Tensor* filenames_tensor;
+    OP_REQUIRES_OK(ctx, ctx->input("filenames", &filenames_tensor));
+    OP_REQUIRES(
+        ctx, filenames_tensor->dims() <= 1,
+        errors::InvalidArgument("`filenames` must be a scalar or a vector."));
+
+    std::vector<string> filenames;
+    filenames.reserve(filenames_tensor->NumElements());
+    for (int i = 0; i < filenames_tensor->NumElements(); ++i) {
+      filenames.push_back(filenames_tensor->flat<string>()(i));
+    }
+
+    string schema;
+    OP_REQUIRES_OK(ctx, ParseScalarArgument<string>(ctx, "schema", &schema));
+
+    int64 buffer_size = -1;
+    OP_REQUIRES_OK(
+        ctx, ParseScalarArgument<int64>(ctx, "buffer_size", &buffer_size));
+    OP_REQUIRES(ctx, buffer_size >= 256,
+                errors::InvalidArgument("`buffer_size` must be >= 256 B"));
+
+    *output = new Dataset(std::move(filenames), schema, buffer_size);
+  }
+
+ private:
+  class Dataset : public DatasetBase {
+   public:
+    explicit Dataset(std::vector<string> filenames, const string& schema,
+                     int64 buffer_size)
+        : filenames_(std::move(filenames)),
+          options_(AvroReaderOptions::CreateAvroReaderOptions(schema,
+                                                              buffer_size)) {}
+
+    std::unique_ptr<IteratorBase> MakeIteratorInternal(const string& prefix) const
+        override {
+      return std::unique_ptr<IteratorBase>(
+          new Iterator({this, strings::StrCat(prefix, "::AvroRecord")}));
+    }
+
+    const DataTypeVector& output_dtypes() const override {
+      static DataTypeVector* dtypes = new DataTypeVector({DT_STRING});
+      return *dtypes;
+    }
+
+    const std::vector<PartialTensorShape>& output_shapes() const override {
+      static std::vector<PartialTensorShape>* shapes =
+          new std::vector<PartialTensorShape>({{}});
+      return *shapes;
+    }
+
+    string DebugString() const override { return "AvroRecordDatasetOp::Dataset"; }
+
+   private:
+    class Iterator : public DatasetIterator<Dataset> {
+     public:
+      explicit Iterator(const Params& params)
+          : DatasetIterator<Dataset>(params) {}
+
+      Status GetNextInternal(IteratorContext* ctx,
+                             std::vector<Tensor>* out_tensors,
+                             bool* end_of_sequence) override {
+        mutex_lock l(mu_);
+        do { // What is the point of this loop???
```

Comment:
This is supposed to loop until a valid record is produced, all the files are read, or there's an error. Also can you turn it into a regular while loop instead of a do-while?